Source/Core/Core/HW/WiiSave.cpp

```diff
@@ -41,525 +48,509 @@ constexpr Md5 s_md5_blanker{{0x0E, 0x65, 0x37, 0x81, 0x99, 0xBE, 0x45, 0x17, 0xA
+      return false;
 
-      if (!raw_save_file.ReadBytes(file_data.data(), file_size))
+    for (const SaveFile& save_file : files)
+    {
+      FileHDR file_hdr{};
+      file_hdr.magic = FILE_HDR_MAGIC;
+      file_hdr.permissions = save_file.mode;
+      file_hdr.attrib = save_file.attributes;
+      file_hdr.type = save_file.type;
+      if (save_file.path.length() > 0x44)
+        return false;
+      std::strncpy(file_hdr.name.data(), save_file.path.data(), file_hdr.name.size());
+
+      std::optional<std::vector<u8>> data;
+      if (file_hdr.type == 1)
       {
-        ERROR_LOG(CONSOLE, "Failed to read data from file: %s", m_files_list[i].c_str());
-        m_valid = false;
+        data = *save_file.data;
+        if (!data)
+          return false;
+        file_hdr.size = static_cast<u32>(data->size());
       }
 
-      m_ios.GetIOSC().Encrypt(IOS::HLE::IOSC::HANDLE_SD_KEY, file_hdr_tmp.iv.data(),
-                              file_data.data(), file_size_rounded, file_data_enc.data(),
-                              IOS::PID_ES);
+      if (!m_file.WriteArray(&file_hdr, 1))
+        return false;
 
-      File::IOFile fpData_bin(m_encrypted_save_path, "ab");
-      if (!fpData_bin.WriteBytes(file_data_enc.data(), file_size_rounded))
+      if (data)
       {
-        ERROR_LOG(CONSOLE, "Failed to write data to file: %s", m_encrypted_save_path.c_str());
+        std::vector<u8> file_data_enc(Common::AlignUp(data->size(), BLOCK_SZ));
+        std::copy(data->cbegin(), data->cend(), file_data_enc.begin());
+        m_iosc.Encrypt(IOS::HLE::IOSC::HANDLE_SD_KEY, file_hdr.iv.data(), file_data_enc.data(),
+                       file_data_enc.size(), file_data_enc.data(), IOS::PID_ES);
+        if (!m_file.WriteBytes(file_data_enc.data(), file_data_enc.size()))
+          return false;
       }
     }
-  }
-}
-
-void WiiSave::do_sig()
-{
-  if (!m_valid)
-    return;
 
-  File::IOFile data_file(m_encrypted_save_path, "rb");
-  if (!data_file)
-  {
-    m_valid = false;
-    return;
+    if (!WriteSignatures())
+    {
+      ERROR_LOG(CORE, "WiiSave::WriteFiles: Failed to write signatures");
+      return false;
+    }
+    return true;
   }
 
-  // Read data to sign.
-  const u32 data_size = m_bk_hdr.size_of_files + 0x80;
-  auto data = std::make_unique<u8[]>(data_size);
-  data_file.Seek(0xf0c0, SEEK_SET);
-  if (!data_file.ReadBytes(data.get(), data_size))
+private:
+  bool WriteSignatures()
   {
-    m_valid = false;
-    return;
-  }
+    const std::optional<BkHeader> bk_header = ReadBkHeader();
+    if (!bk_header)
+      return false;
 
-  // Sign the data.
-  IOS::CertECC ap_cert;
-  Common::ec::Signature ap_sig;
-  m_ios.GetIOSC().Sign(ap_sig.data(), reinterpret_cast<u8*>(&ap_cert), Titles::SYSTEM_MENU,
-                       data.get(), data_size);
+    // Read data to sign.
+    const u32 data_size = bk_header->size_of_files + 0x80;
+    auto data = std::make_unique<u8[]>(data_size);
+    m_file.Seek(0xf0c0, SEEK_SET);
```

Comment:
I don't suppose we have constants or appropriate `sizeof`'s/`offsetof`'s for `0x80` and `0xf0c0`?

Reply:
Yeah, we don't and I'm not sure what these correspond to :(

Reply:
Oh, actually I think I found what these are. 0x80 is probably `sizeof(BkHeader)` and 0xf0c0 `sizeof(Header)`. I've pushed another commit to replace these magic numbers (and added some static_assert to make sure we don't mess up the structs)